### PR TITLE
Add strnstr implementation for non-BSD

### PR DESCRIPTION
**Pull request by lherman-cs**
`STRNSTR` is defined in common def. But, unfortunately, `strnstr` is only defined in BSD. So, to make `STRNSTR` available for all platforms, this PR will add a default implementation in Rabin Karp algorithm.

`STRNSTR` is also needed for this PR, https://github.com/awslabs/amazon-kinesis-video-streams-webrtc-sdk-c/pull/600.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/utils/tst/StringSearch.cpp [hunk 1]
```diff
@@ -0,0 +1,60 @@
+#include "UtilTestFixture.h"
+
+class StringSearchFunctionalityTest : public UtilTestBase {
+};
+
+TEST_F(StringSearchFunctionalityTest, NegativeInvalidInput) {
```
Comment (MushMal) [inline src/utils/tst/StringSearch.cpp:6]: Add some tests for empty. Also, add cases where needle is larger than the stack

File: src/utils/src/String.c [hunk 2]
```diff
@@ -449,4 +449,66 @@ STATUS tolowerupperstr(PCHAR pStr, UINT32 strLen, BOOL toUpper, PCHAR pConverted
...
+PCHAR defaultStringStrnstr(PCHAR haystack, PCHAR needle, SIZE_T len)
+{
+    UINT32 prime = 16777619;
+    UINT32 windowHash = 0;
+    UINT32 needleHash = 0;
+    UINT32 i;
+    UINT32 square = prime, power = 1;
+    UINT32 haystackSize, needleSize;
+
+    if (needle == NULL) {
```
[the remark this answers is not in the document]
Comment (lherman-cs) [reply src/utils/src/String.c:468]: No need, since no loops will be executed and will go straight to the final return, `return NULL;`

File: src/utils/src/String.c [hunk 3]
```diff
@@ -449,4 +449,66 @@ STATUS tolowerupperstr(PCHAR pStr, UINT32 strLen, BOOL toUpper, PCHAR pConverted
...
+        windowHash = windowHash * prime + (UINT32) haystack[i];
+        needleHash = needleHash * prime + (UINT32) needle[i];
+    }
+
+    if (windowHash == needleHash && STRNCMP(haystack, needle, needleSize) == 0) {
+        return haystack;
+    }
+
+    // find the largest power in O(log n)
+    for (i = needleSize; i > 0; i /= 2) {
```
[the remark this answers is not in the document]
Comment (lherman-cs) [reply src/utils/src/String.c]: added a small explanation

File: src/utils/tst/StringSearch.cpp (hunk 1 above)
Comment (lherman-cs) [reply src/utils/tst/StringSearch.cpp:6]: Makes sense! I've added the unit tests.

File: src/common/include/com/amazonaws/kinesis/video/common/CommonDefs.h [hunk 4]
```diff
@@ -575,6 +575,21 @@ typedef struct stat STAT_STRUCT;
 
 #endif
 
+//
```
[the remark this answers is not in the document]
Comment (lherman-cs) [reply src/common/include/com/amazonaws/kinesis/video/common/CommonDefs.h:578]: ah makes sense.

File: src/utils/src/String.c [hunk 5]
```diff
@@ -449,4 +449,89 @@ STATUS tolowerupperstr(PCHAR pStr, UINT32 strLen, BOOL toUpper, PCHAR pConverted
...
+        return haystack;
+    }
+
+    // Precompute the largest power in O(log n). 
+    //
+    // This algorithm takes an advantage for the fact that a number that's multiplied 
+    // by itself will double the power.
+    //
+    // For example:
+    //    a ^ 4 = (a ^ 2) * (a ^ 2) = a * a * a * a * a * a * a * a
```
Comment (MushMal) [inline src/utils/src/String.c:497]: You have a ^ 8 in  a * a * a * a * a * a * a * a

Comment (lherman-cs) [reply src/utils/src/String.c:497]: Oops, I can't count...
